clusterloader2/pkg/measurement/common/generic_query_measurement.go

```diff
@@ -104,38 +105,64 @@ func (g *genericQueryGatherer) IsEnabled(config *measurement.Config) bool {
 	return true
 }
 
+func key(metric model.Metric, dimensions []string) (string, map[string]string) {
+	s := make([]string, 0)
+	m := make(map[string]string)
+	for _, dimension := range dimensions {
+		val := string(metric[model.LabelName(dimension)])
+		s = append(s, val)
+		m[dimension] = val
+	}
+	return fmt.Sprintf("%v", s), m
+}
+
 func (g *genericQueryGatherer) Gather(executor QueryExecutor, startTime, endTime time.Time, config *measurement.Config) ([]measurement.Summary, error) {
 	var errs []error
-	data := map[string]float64{}
+	dataItems := map[string]*measurementutil.DataItem{}
 	for _, q := range g.Queries {
 		samples, err := g.query(q, executor, startTime, endTime)
 		if err != nil {
 			return nil, err
 		}
 
-		if len(samples) > 1 {
-			errs = append(errs, errors.NewMetricViolationError(q.Name, fmt.Sprintf("too many samples: query returned %v streams, expected 1", len(samples))))
-		}
 		if len(samples) == 0 {
 			klog.Warningf("query returned no samples for %v: %v", g.MetricName, q.Name)
 			continue
 		}
 
-		val := float64(samples[0].Value)
-
-		thresholdMsg := "none"
-		if q.Threshold != nil {
-			thresholdMsg = fmt.Sprintf("%v", *q.Threshold)
-		}
-		klog.V(2).Infof("metric: %v: %v, value: %v, threshold: %v", g.MetricName, q.Name, val, thresholdMsg)
-
-		if q.Threshold != nil && val > *q.Threshold {
-			errs = append(errs, errors.NewMetricViolationError(q.Name, fmt.Sprintf("sample above threshold: want: less or equal than %v, got: %v", q.Threshold, val)))
+		for _, sample := range samples {
+			k, labels := key(sample.Metric, g.Dimensions)
+
+			dataItem, ok := dataItems[k]
+			if !ok {
+				dataItem = &measurementutil.DataItem{
+					Data:   make(map[string]float64),
+					Unit:   g.Unit,
+					Labels: labels,
+				}
+				dataItem.Labels["MetricName"] = g.MetricName
+				dataItems[k] = dataItem
+			}
+
+			val := float64(sample.Value)
+			_, ok = dataItem.Data[q.Name]
+			if ok {
+				errs = append(errs, errors.NewMetricViolationError(q.Name, fmt.Sprintf("too many samples for %s: query returned %v streams, expected 1", k, len(samples))))
```

Comment:
I'm not sure that len(samples) is the number of streams with the same key. Could you check?

Reply:
Done